### PR TITLE
Single NUMA node policy unittest

### PR DESCRIPTION
**Pull request by cynepco3hahue**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 1]
```diff
@@ -0,0 +1,1044 @@
...
+	for _, testCase := range testCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			p := NewPolicySingleNUMA(testCase.machineInfo, testCase.reserved, topologymanager.NewFakeManager())
+			s := state.NewMemoryState()
+			s.SetMachineState(testCase.machineState)
+			s.SetMemoryAssignments(testCase.assignments)
+
+			err := p.Allocate(s, testCase.pod, &testCase.pod.Spec.Containers[0])
+			if !reflect.DeepEqual(err, testCase.expectedError) {
+				t.Errorf("The actual error %v different from the expected one %v", err, testCase.expectedError)
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:751]: `error %v differs` of `error %v is different`

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 2]
```diff
@@ -0,0 +1,1044 @@
...
+	for _, testCase := range testCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			p := NewPolicySingleNUMA(testCase.machineInfo, testCase.reserved, topologymanager.NewFakeManager())
+			s := state.NewMemoryState()
+			s.SetMachineState(testCase.machineState)
+			s.SetMemoryAssignments(testCase.assignments)
+
+			err := p.Start(s)
+			if !reflect.DeepEqual(err, testCase.expectedError) {
+				t.Errorf("The actual error: %v is different from the expected one: %v", err, testCase.expectedError)
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:543]: as below

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 3]
```diff
@@ -0,0 +1,1044 @@
...
+							{
+								// size in KB
+								PageSize: pageSize1Gb,
+								NumPages: 4,
+							},
+						},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("[memorymanager] hugepages-1Gi reserved by containers different from the machine state reserved"),
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:530]: `differ` or `are different`

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 4]
```diff
@@ -0,0 +1,1044 @@
...
+					"container1": {
+						{
+							NUMAAffinity: 0,
+							Type:         v1.ResourceMemory,
+							Size:         512 * mb,
+						},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("[memorymanager] machine state can not be empty, when it has memory assignments"),
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:143]: I think comma is not necessary here.

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 5]
```diff
@@ -0,0 +1,1044 @@
...
+								PageSize: pageSize1Gb,
+								NumPages: 1,
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			description: "should fail when machine state does not have all NUMA nodes",
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:188]: :+1:

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 6]
```diff
@@ -0,0 +1,1044 @@
...
+								NumPages: 1,
+							},
+						},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("[memorymanager] machine state has different size of the system reserved memory"),
+		},
+		{
+			description: "should fail when machine state reserved memory is different from memory of all containers memory assignments",
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:341]: `from the memory of`

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 7]
```diff
@@ -0,0 +1,1044 @@
...
+							{
+								// size in KB
+								PageSize: pageSize1Gb,
+								NumPages: 1,
+							},
+						},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("[memorymanager] memory reserved by containers different from the machine state reserved"),
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:384]: `is different` or `differs`

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 8]
```diff
@@ -0,0 +1,1044 @@
...
+								NumPages: 2,
+							},
+						},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("[memorymanager] machine state has different size of the system reserved hugepages-1Gi"),
+		},
+		{
+			description: "should fail when machine state reserved hugepages is different from hugepages of all containers memory assignments",
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:471]: `the hugepages reserved machine state`
`the hugepages of`

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 9]
```diff
@@ -0,0 +1,1044 @@
...
+	if len(mb1) != len(mb2) {
+		return false
+	}
+
+	copyMemoryBlocks := make([]state.Block, len(mb2))
+	copy(copyMemoryBlocks, mb2)
+	for _, block := range mb1 {
+		for i, copyBlock := range copyMemoryBlocks {
+			if reflect.DeepEqual(block, copyBlock) {
+				// move the element that equal to the block to the end of the slice
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:78]: `is equal` or `equals` ?

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 10]
```diff
@@ -0,0 +1,1044 @@
...
+		},
+		Requests: v1.ResourceList{
+			v1.ResourceCPU:    resource.MustParse("1000Mi"),
+			v1.ResourceMemory: resource.MustParse("1Gi"),
+			hugepages1Gi:      resource.MustParse("1Gi"),
+		},
+	}
+)
+
+func isMemoryBlocksEqual(mb1, mb2 []state.Block) bool {
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:68]: `areMemoryBlocksEqual` ?

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 11]
```diff
@@ -0,0 +1,1044 @@
...
+
+				break
+			}
+		}
+	}
+
+	return len(copyMemoryBlocks) == 0
+}
+
+func isContainerMemoryAssignmentsEqual(cma1, cma2 state.ContainerMemoryAssignments) bool {
```
Comment (cezaryzukowski) [inline pkg/kubelet/cm/memorymanager/policy_single_numa_test.go:92]: `areContainer...`?

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 12]
```diff
@@ -0,0 +1,1044 @@
...
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			p := NewPolicySingleNUMA(testCase.machineInfo, testCase.reserved, topologymanager.NewFakeManager())
+			s := state.NewMemoryState()
+			s.SetMachineState(testCase.machineState)
+			s.SetMemoryAssignments(testCase.assignments)
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/memorymanager/policy_single_numa_test.go]: done

File: pkg/kubelet/cm/memorymanager/policy_single_numa_test.go [hunk 13]
```diff
@@ -0,0 +1,1044 @@
...
+						NUMANodeAffinity: affinity0,
+						Preferred:        true,
+					},
+				},
+			},
+		},
+		{
+			description: "should calculate new topology hints, when the container does not exist under assignments",
+			assignments: state.ContainerMemoryAssignments{},
+			expectedAssignments: state.ContainerMemoryAssignments{
```
[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/memorymanager/policy_single_numa_test.go]: good catch, done

**Pull request merged**